bliss/simulator/background.py

```diff
@@ -1,48 +1,64 @@
-from typing import Tuple
+from typing import Dict, Tuple
 
 import numpy as np
 import torch
 from einops import rearrange
+from omegaconf import OmegaConf
+from omegaconf.dictconfig import DictConfig
 from torch import Tensor, nn
 
 # ideally this module wouldn't depend on any specific survey
 from bliss.surveys.sdss import SloanDigitalSkySurvey
 
 
 class ConstantBackground(nn.Module):
-    def __init__(self, background: Tuple[float, ...]):
+    def __init__(self, background: Tuple[float, ...], **_kwargs):
         super().__init__()
-        bg: Tensor = torch.tensor(background)
-        bg = rearrange(bg, "c -> 1 c 1 1")
+        bg: Tensor = torch.tensor([background])
+        bg = rearrange(bg[0], "c -> 1 c 1 1")
         self.register_buffer("background", bg, persistent=False)
 
-    def sample(self, shape) -> Tensor:
+    def sample(self, shape, **_kwargs) -> Tensor:
         assert isinstance(self.background, Tensor)
         batch_size, c, hlen, wlen = shape
         return self.background.expand(batch_size, c, hlen, wlen)
 
 
 class SimulatedSDSSBackground(nn.Module):
-    def __init__(self, sdss_dir, run, camcol, field, bands):
+    def __init__(self, sdss_fields: DictConfig):
         super().__init__()
-        sdss_data = SloanDigitalSkySurvey(
-            sdss_dir=sdss_dir,
-            run=run,
-            camcol=camcol,
-            fields=(field,),
-            bands=bands,
-        )
-        background = torch.from_numpy(sdss_data[0]["background"])
-        background = rearrange(background, "c h w -> 1 c h w", c=len(bands))
+        sdss_dir = sdss_fields["dir"]
+        field_list = sdss_fields["field_list"]
+        bands = sdss_fields["bands"]  # use same bands across fields
+
+        # Add all backgrounds from specified run/col/field to list
+        backgrounds = []
+        for param_obj in field_list:
+            params: Dict = OmegaConf.to_container(param_obj)  # type: ignore
+            sdss = SloanDigitalSkySurvey(sdss_dir=sdss_dir, bands=bands, **params)
+            backgrounds.extend([field["background"] for field in sdss])  # type: ignore
+
+        background = torch.from_numpy(np.stack(backgrounds, axis=0))
+        if len(background.size()) == 3:  # convert to 4d tensor if only one background
```

Comment:
If there's only one background, can we require that background is already a 4d tensor with a singleton first dimension? (It's nice to avoid conditional expressions when possible.)

Reply:
Oh I see, background isn't provided as an argument to this function. Still, is there a way to avoid treating 1 background as a special case?

Reply:
That's what the next line is doing! :) Before there was only one background so this was always executed, now since we might have multiple backgrounds already we need to check first.
https://github.com/prob-ml/bliss/blob/95e3087cf69e40da628e33060245e09b851d1d5e/bliss/simulator/background.py#L42-L43

Reply:
I was hoping to avoid even this one-line special case, but perhaps that's not possible (?) Why does the dimension of background vary here?

Reply:
Is this if condition ever true now? Even if you call `np.stack` with a singleton list, it still adds a dimension:
```
>>> two_d_array = np.ones((4,5))
>>> np.stack([two_d_array,], axis=0).shape
(1, 4, 5)
```

Reply:
Huh you're right, it already is 1 x c x h x w for a single background already, good catch! We can just remove this check then.
![image](https://github.com/prob-ml/bliss/assets/26699876/36f197a9-6e79-45ee-96d3-d4b7e2f0ee45)

Reply:
Fixed